src/attributes.jl

```diff
@@ -188,45 +193,82 @@ function h5readattr(filename, name::AbstractString)
 end
 
 
-struct Attributes
-    parent::Union{File,Object}
+struct AttributeDict <: AbstractDict{String,Any}
```

Comment:
Let's consider parameterizing this so that we can have a more specific type than `Any`, which can be the default is not specified.

Reply:
I don't see how that could be made to work, or really,what advantage it would give. Can you expand on what you had in mind? What would happen if you attempted to read or write an attribute of a different type?

My intention was to provide a simple, easy-to-use interface. I think if you want to specify the return type, it would be better to do in the mid-level interface, e.g. adding a return type to `read_attribute`.

Reply:
Perhaps `AttributeDict{T}` could use `convert(T, ...)`? If `convert` throws an error, then so be it.

Reply:
It adds a lot of complexity and potential brittleness, and I'm still skeptical as to why it would be helpful? The files I've seen with attributes use a mix of different types (typically strings, integers and vectors of integers).

Reply:
Here's my prototype diff:
https://github.com/JaneliaSciComp/HDF5.jl/commit/c3a5db33b1cf96e245c6ae8bf5edaf61a10073e6
I added a `try` to `Base.iterate` to skip over items where `convert` fails.

Type stability can be helpful. Since I'm accessing the HDF5 file in Julia rather than Python, I would like to take advantage of the type system. In many cases, I may have prior knowledge of the type I'm expecting the attribute to be, or I can glean that from the API. This provides an easy way to enforce knowledge of the type, allowing subsequent code to be type stable without having to use a completely different API. It also throws an error if my assumptions are violated.

Having the parameter default to `Any` produces the exact same results you have now. Allowing for a parameter allows the same API to be extended to encourage type stability.


```julia
julia> attrs(h5f)
AttributeDict of HDF5.Group: / (file: test.h5) with 4 attributes:
  "bye"     => "3.0"
  "hello"   => 5
  "someint" => 3
  "test"    => 5.5

julia> attrs(h5f, Int)
AttributeDict of HDF5.Group: / (file: test.h5) with 4 attributes:
  "hello"   => 5
  "someint" => 3

julia> attrs(h5f, String)
AttributeDict of HDF5.Group: / (file: test.h5) with 4 attributes:
  "bye" => "3.0"

julia> attrs(h5f, Float64)
AttributeDict of HDF5.Group: / (file: test.h5) with 4 attributes:
  "hello"   => 5.0
  "someint" => 3.0
  "test"    => 5.5

julia> attrs(h5f, Int)["test"]
ERROR: InexactError: Int64(5.5)
Stacktrace:
 [1] Int64
   @ ./float.jl:812 [inlined]
 [2] convert(#unused#::Type{Int64}, x::Float64)
   @ Base ./number.jl:7
 [3] getindex(x::HDF5.AttributeDict{Int64}, name::String)
   @ HDF5 ~/.julia/dev/HDF5/src/attributes.jl:224
 [4] top-level scope
   @ REPL[12]:1
```

There a few things to explore here, such as passing the parameter as the `memtype` to `read_attribute`, but I think this would allow this API to go a lot further. Since we are talking about parameterizing `Dataset` as well, I think adding a type parameter to `AttributeDict` would be consistent with future direction of this package.

Reply:
`Dataset` would be `struct Dataset{T,N} <: AbstractArray{T,N}` or `struct Dataset{T,N} <: AbstractDiskArray{T,N}`.

I wanted to explore if `Attribute` could become `Attribute{T}` and have a corresponding `struct Attributes{T2} <: AbstractDict{String, Attribute{<:T2}}`. However, you want to deprecate that interface, which is why I'm exploring the idea here.

I would be fine if we parameterized the struct but only implemented methods for `AttributeDict{Any}` in this pull request.

Reply:
I still don't see the benefit of adding a parameter to `AttributeDict`: how would it help?